### DISCUSSION TITLE
Enums are not dropped when using `drop_all` after adding table using `Table.to_metadata`

### DISCUSSION THREAD

**Discussion by BraedonLeonard** [sqlalchemy/sqlalchemy#11798] [Usage Questions]
I noticed some unexpected behaviour when attempting to use `MetaData.drop_all` to drop all tables and types. I found a minimal reproducible case, detailed below.

First, we setup 2 metadata objects, one which we use when first initializing the table, and one which we later add the table to using `Table.to_metadata`:
```python
import enum
from sqlalchemy import Column, Enum, Integer, MetaData, Table, create_engine

class MyEnum(enum.StrEnum):
    value_1 = enum.auto()
    value_2 = enum.auto()

metadata1 = MetaData()

my_table = Table(
    "my_table",
    metadata1,
    Column("id", Integer(), primary_key=True),
    Column("coalesce", Enum(MyEnum, metadata=metadata1), nullable=False),
)

metadata2 = MetaData()
my_table.to_metadata(metadata2)
```

Once we've done that, we can create an engine pointing at a local postgres instance. For simplicity, I just ran the following `docker run` to start an ephemeral postgres DB:
```sh
docker run -p 5432:5432 -e POSTGRES_PASSWORD=mysecretpassword postgres
```

For the engine:
```python
engine = create_engine("postgresql://postgres:mysecretpassword@localhost:5432")
```

At this point, we are able to sanity check that `metadata1.create_all` and `metadata.drop_all` are working as expected:
```python
# `my_table` and `myenum` are created
metadata1.create_all(engine)
# `my_table` and `myenum` are dropped
metadata1.drop_all(engine)
```

After that, we can do the same with `metadata2` and see that it drops the table, but not the type:
```python
# `my_table` and `myenum` are created
metadata2.create_all(engine)
# `my_table` is dropped, but `myenum` is not
metadata2.drop_all(engine)
```

I'm pretty sure this is a bug, but it's possible I'm just misusing these features.

**Comment by zzzeek** [comment]
nah this one is weird, because it's *only* the "drop" that doesnt happen.  it's also dependent on the Enum being attached to metadata1 as you've done.

there's a transfer of events and such that has to take place for the enum type here and it would be strange if I made the transfer work for "create" but not for "drop" and then only if the enum were associated with an existing metadata, so this feels like it might be trying to accommodate some other use case, but it would be better if it behaved consistently so let me see where this happens.

**Reply by zzzeek** [reply]
OK it makes "sense", there's a reason it happens but ultimately there's a likely bug that leads it into this condition.   the metadata assigned to the Enum is not set to the new one, so I want to fix that.   I can see this being a bit of a circus for applications or test suites that are already working around this one but I can't put everything into 2.1 so here goes

**Reply by zzzeek** [reply]
I wish github would automatically tag this, this is in #11802 for the next release

**Reply by CaselIT** [reply]
> I wish github would automatically tag this, this is in #11802 for the next release

it does, but it kinda hides it here
![image](https://github.com/user-attachments/assets/55624dcb-7821-460d-821e-9a6402c0809b)
